docs/proposal/20250224-rosa-networking.md

```diff
@@ -0,0 +1,221 @@
+
+Adopting the CloudFormation template used by rosa-cli would mean that CAPA and the `RosaNetwork` custom resource would be relying on a mechanism that is know to work well and any changes or fixes implemented in ROSA CLI would be picked up automatically in CAPA.
+
+In practical terms, implementation of the proposal would mean:
+1. A new namespaced custom resource definition `RosaNetwork` in CAPA with five attributes: name, AZ count, list of availability zones, region and CIDR block for VPC. `availabilityZoneCount`, `availabilityZones`, `region` and `cidrBlock` will become the `spec` part of the new `RosaNetwork` type, name of the cloudformation stack will be the same as `metadata.name`.
+
+   `RosaNetwork` spec example:
+    ```
+    kind: RosaNetwork
+    metadata:
+      name: rosa-network-01
+      namespace: default
+    spec:
+      availabilityZoneCount: 3
+      region: us-west-2
+      cidrBlock: 10.0.0.0/16
+    ```
+
+   `RosaNetwork` spec example with specified availability zones:
+    ```
+    kind: RosaNetwork
+    metadata:
+      name: rosa-network-01
+      namespace: default
+    spec:
+      availabilityZones:
+      - us-west-2a
+      - us-west-2d
+      region: us-west-2
+      cidrBlock: 10.0.0.0/16
+    ```
+
+1. A new reconciler for the new custom resource, implementing creation and deletion. The reconciler will be using an existing [CloudFormation template from ROSA CLI](https://github.com/openshift/rosa/blob/master/cmd/create/network/templates/rosa-quickstart-default-vpc/cloudformation.yaml) and will use [AWS CloudFormation API](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/cloudformation) to [create](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/cloudformation#Client.CreateStack) and [delete](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/cloudformation#Client.DeleteStack) the AWS CloudFormation stack.
+
+   Outputs and resources created in the cloudformation stack will be tracked under `status` of the `RosaNetwork` type. In particular, the `status` will contain the list of public and private subnets and availability zones, grouped together by the availability zones.
+
+   Example:
+   ```
+   kind: RosaNetwork
+   metadata:
+     name: rosa-network-01
+     namespace: default
+   status:
+     subnets:
+     - availabilityZone: us-west-2a
+       publicSubnet: subnet-0d9f28ba991b93514
+       privateSubnet: subnet-1d9f28ba992a83514
+     - availabilityZone: us-west-2b
+       publicSubnet: subnet-2d7f18c09f1b43512
+       privateSubnet: subnet-2d7f58c09f1b43512
+     - availabilityZone: us-west-2c
+       publicSubnet: subnet-1d7e19c0af1c4c57f
+       privateSubnet: subnet-7d7e19c0af1f4d57f
+   ```
+
+   All resources created in the cloudformation stack will be tracked under `status.resources` array:
```

Comment:
What's the reason for setting `subnets` as a top-level field instead of listing them under `status.resources.subnets`?

Reply:
@nrb in the end the subnets would be listed under both `status.subnets` (grouped by AZ) and `status.resources`. The `status.resources` will be a list filled with the data propagated from AWS as the stack creation progresses. My thought here was to duplicate the results we're interested in (subnets & AZs) so that we don't have to parse it out of the `status.resources` (i.e. a convenience function).

Reply:
So if I could rephrase to confirm I understand - the subnets are promoted to a top level resource primarily for ease of readability?

Reply:
@nrb Yes, I guess you could say that. Although I meant it mostly for the benefit of readibility by other controllers: AZs and subnet ids is what the ROSA controller needs to provision the actual cluster.

Do you think this is not a good idea?